sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java

```diff
@@ -224,9 +224,13 @@ public ColumnarBatch resultBatch() {
 
   /**
    * Can be called before any rows are returned to enable returning columnar batches directly.
+   *
+   * @param maxColumns The maximum number of fields supported by whole stage codegen.
    */
-  public void enableReturningBatches() {
-    returnColumnarBatch = true;
+  public boolean tryEnableReturningBatches(int maxColumns) {
```

Comment:
Yes, we should do this in planner, by checking the schema, which requires some cleanup in parquet reader, currently which still use runtime exception to do the switch. cc @nongli @sameeragarwal 

Should we wait for that?